Implementation of CDS writer?

keflavich opened the pull request:
I think this is a reasonable attempt at a CDS table writer.  It's missing some key things:
-can columns have "explanation" attributes?  Or, how do we fill the "explanation" part of the CDS header?
-Users may want to write out CDS tables as separate header/other files; this is not allowed right now

I've included a simple test that looks a touch silly.  It would probably be good to make the write command somewhat more user-accessible, but that's a change on par with overloading the record array so might not be practical.

astr2600student committed:
implemented the CDS writer (tentatively)

astr2600student committed:
I didn't really understand how git worked... NOW I think I've added the
CDS writer

taldcroft commented:
Looks like a good start.  For the explanation attributes you could add a "descriptions" keyword arg to the class (see Latex again for examples).  "descriptions" could be a dict with a key/val pair for each column with a description.  Alternatively a list of explanations corresponding to each column could work, but this can get tedious if there are lots of cols and you just have a few with explanations.

It would certainly be possible to add an attribute "description" to the Column class.  Various readers would need to populate this attribute if possible.

Having some real tests would be good, along with updated documentation showing an example or two and listing known limitations.  Following some of the PEP8 style guidelines would help the code readability, especially space before & after most operators, space after comma, etc.

hamogu commented:
I am really looking forward to see your CDS writer in action, that looks very promising.

If you decide to input via a dictionary and it works, maybe we should use that approach for all writer classes in the future, which require extra input in some form.

Commit:
added some documentation to CDS writer, along with support for the descr
header attribute.  Currently only CDS readers fill the descr attribute;
IPAC tables have descriptions but the reader currently does not support
them

Commit:
Added more documentation to the CDS writer along with more spacing
prettification.  The CDS writer test case now uses a somewhat more
intricate input.  However, testing the "descr" writer still requires a
reasonable "descr" reader, which isn't implemented for the IPAC case.

Commit:
Added test case for CDS -> CDS (which tests descr)
multi-line descriptions are now allowed, but note that they aren't read
exactly right at the moment - they should be space-separated (or at
least whitespace-separated) but are not when they are read in.  Also, it
might make sense to try to split 'descr' along spaces, but the current
method is more straightforward

keflavich commented:
The "descr" keyword already exists for the CDS class, but not for the IPAC class (which should also support descriptions; see e.g. http://irsa.ipac.caltech.edu/data/BOLOCAM_GPS/tables/bolocam_gps_v1_0_1.tbl).  I think that's the next thing on the To Do list.

I've added some documentation, an implementation of "descr" writing that checks for existence first, and a slightly more complex set of test cases with comments.  I've added some PEP8 compliance stuff, but I'm probably not fully compliant (I still think of " "*8 as an 8-string character... a single term).

taldcroft commented:
What specification of the CDS format are you referencing for the write function?  I'm looking at http://vizier.u-strasbg.fr/doc/catstd-3.1.htx and the CDS writer output does not comply on several counts (see in particular point #9).  The output of the CDS writer cannot be read by the asciitable CDS reader unless I am doing something wrong (e.g. t/cds_convert_to_cds.cds).

The tests need to be meaningful by writing a table and then comparing some or all of that table to the expected answer.  Also try reading back the file using the asciitable.Cds reader.  Tests should write the table to a StringIO object instead of an actual file, or else use files in a temporary directory that gets automatically cleaned up even in the event of an exception during test.

keflavich commented:
I ran my test successfully before the last commit, but am now running into some serious errors, the biggest being the variable table types returned by the reader - in one case, it is a <class 'numpy.core.records.record'>.  In another, it's <class 'numpy.ma.core.MaskedArray'>.  Unfortunately, it looks like these behave very differently, and reshaping the arrays into tuples isn't straightforward.  I'll continue work on this slowly, but it has proved to be a greater challenge than I had originally appreciated.
